File: loom/src/loom/util/stable_id.c
// Copyright 2026 The IREE Authors
//
// Licensed under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

#include "loom/util/stable_id.h"

uint64_t loom_stable_id_from_string(iree_string_view_t key) {
  uint64_t value = UINT64_C(0xCBF29CE484222325);
  for (iree_host_size_t i = 0; i < key.size; ++i) {
    value ^= (uint8_t)key.data[i];
    value *= UINT64_C(0x100000001B3);
  }
  value &= UINT64_C(0x7FFFFFFFFFFFFFFF);
  return value == LOOM_STABLE_ID_NONE ? UINT64_C(1) : value;
}

File: loom/py/loom/stable_id.py
# Copyright 2026 The IREE Authors
#
# Licensed under the Apache License v2.0 with LLVM Exceptions.
# See https://llvm.org/LICENSE.txt for license information.
# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

"""Stable integer identities derived from textual keys."""

from __future__ import annotations

STABLE_ID_NONE = 0


def stable_id_from_string(key: str) -> int:
    """Returns a stable non-zero 63-bit identity for ``key``."""

    value = 0xCBF29CE484222325
    for byte in key.encode("utf-8"):
        value ^= byte
        value = (value * 0x100000001B3) & 0xFFFFFFFFFFFFFFFF
    value &= 0x7FFFFFFFFFFFFFFF
    return 1 if value == STABLE_ID_NONE else value
